homeassistant/components/tado/device_tracker.py

```diff
@@ -1,149 +1,141 @@
+    """A Tado Device Tracker entity."""
+
+    def __init__(
+        self,
+        device_id: str,
+        device_name: str,
+        tado: Any,
+    ) -> None:
+        """Initialize a Tado Device Tracker entity."""
+        super().__init__()
+        self._device_id = device_id
+        self._device_name = device_name
+        self._tado = tado
+        self._active = False
+
+    @callback
+    def update_state(self) -> None:
+        """Update the Tado device."""
+        _LOGGER.debug(
+            "Updating Tado mobile device: %s (ID: %s)",
+            self._device_name,
+            self._device_id,
+        )
+        device = self._tado.data["mobile_device"][self._device_id]
 
-        # If there's a home_id, we need a different API URL
-        if self.home_id is None:
-            self.tadoapiurl = "https://my.tado.com/api/v2/me"
+        self._active = False
+        if device.get("location") is not None and device["location"]["atHome"]:
+            _LOGGER.debug("Tado device %s is at home", device["name"])
+            self._active = True
         else:
-            self.tadoapiurl = "https://my.tado.com/api/v2/homes/{home_id}/mobileDevices"
-
-        # The API URL always needs a username and password
-        self.tadoapiurl += "?username={username}&password={password}"
-
-        self.websession = None
-
-        self.success_init = asyncio.run_coroutine_threadsafe(
-            self._async_update_info(), hass.loop
-        ).result()
-
-        _LOGGER.info("Scanner initialized")
-
-    async def async_scan_devices(self):
-        """Scan for devices and return a list containing found device ids."""
-        await self._async_update_info()
-        return [device.mac for device in self.last_results]
-
-    async def async_get_device_name(self, device):
-        """Return the name of the given device or None if we don't know."""
-        filter_named = [
-            result.name for result in self.last_results if result.mac == device
-        ]
-
-        if filter_named:
-            return filter_named[0]
-        return None
-
-    @Throttle(MIN_TIME_BETWEEN_SCANS)
-    async def _async_update_info(self):
-        """Query Tado for device marked as at home.
-
-        Returns boolean if scanning successful.
-        """
-        _LOGGER.debug("Requesting Tado")
-
-        if self.websession is None:
-            self.websession = async_create_clientsession(
-                self.hass, cookie_jar=aiohttp.CookieJar(unsafe=True)
+            _LOGGER.debug("Tado device %s is not at home", device["name"])
+
+        self.async_write_ha_state()
+
+    async def async_added_to_hass(self) -> None:
+        """Register state update callback."""
+        _LOGGER.debug("Registering Tado device tracker entity")
+        self.async_on_remove(
+            async_dispatcher_connect(
+                self.hass,
+                SIGNAL_TADO_MOBILE_DEVICE_UPDATE_RECEIVED,
+                self.update_state,
             )
+        )
```

Comment:
Call `self.update_state` once here to set the initial state correctly.

Thinking about this it would actually be good with a separate method to calculate the state and then another method that includes the call to `self.async_write_ha_state`. Then we could just calculate state here since there will always be a state update at the end of entity addition by default.

Reply:
Wouldn't this revert back the principe of:

```
 @callback
    def on_demand_update(self) -> None:
        """Update state on demand."""
        self.update_state()
        self.async_write_ha_state()
```

Reply:
Update provided.